lib/Fs/NodeHelper.php

```diff
@@ -251,4 +260,122 @@ public static function folderHasSubPage(Folder $folder, string $title): int {
 
 		return 0;
 	}
+
+	/**
+	 * Recursively collect text from a CommonMark node and its descendants
+	 */
+	private static function collectText(Node $node): string {
+		if ($node instanceof Text) {
+			return $node->getLiteral();
+		}
+
+		$out = '';
+		foreach ($node->children() as $child) {
+			$out .= self::collectText($child);
+		}
+
+		return $out;
+	}
+
+	/**
+	 * @throws CommonMarkException
+	 */
+	public static function getLinksFromContent(string $content): array {
+		$environment = new Environment();
+		$environment->addExtension(new CommonMarkCoreExtension());
+		$parser = new MarkdownParser($environment);
+		$document = $parser->parse($content);
+		$walker = $document->walker();
+
+		$links = [];
+		while ($event = $walker->next()) {
+			if (! $event->isEntering()) {
+				continue;
+			}
+			$node = $event->getNode();
+			if (!($node instanceof Link)) {
+				continue;
+			}
+
+			$textParts = [];
+			foreach ($node->children() as $child) {
+				$textParts[] = self::collectText($child);
+			}
+			$text = trim(implode('', array_filter($textParts)));
+
+			$links[] = [
+				'text' => $text,
+				'href' => $node->getUrl(),
+				'title' => $node->getTitle() ?? '',
+			];
+		}
+
+		return $links;
+	}
+
+	/**
+	 * @throws CommonMarkException
+	 */
+	public static function getLinkedPageIds(Collective $collective, string $content, array $trustedDomains = []): array {
+		$links = self::getLinksFromContent($content);
+
```

Comment:
I think this would benefit from being split into smaller functions. Something like


```suggestion
	public static function getLinkedPageIds(Collective $collective, string $content, array $trustedDomains = []): array {
		$links = self::getLinksFromContent($content);
		$hrefs = self::getHrefsFromLinks($links, $collective, $trustedDomains);
		$pageIds = self::getPageIdsFromHrefs($hrefs);
		return array_unique($pageIds, SORT_NUMERIC);

```